### ISSUE TITLE
[FR] Allow giving a model name when serving with mlserver

### ISSUE THREAD

**Issue by dingobar** [mlflow/mlflow#6657]
### Willingness to contribute

Yes. I would be willing to contribute this feature with guidance from the MLflow community.

### Proposal Summary

When we serve models with mlserver, it is important to have traceability: Which model gave which predictions? The way it is now, `mlflow models serve <uri> --enable-mlserver` gives a default name that cannot be changed (`mlflow-model`, see https://github.com/mlflow/mlflow/blob/8098785831a3414bed62ec3239b60a6a500f4dbd/mlflow/pyfunc/mlserver.py#L6). It would be nice to have one of these options:

- Ability to give a model name when running `mlflow models serve` or `mlflow models build-docker`, that is used as the model name in the `mlserver` metadata, alternatively the ability to give a name suffix like `mlflow-model-<some uuid for example>`, run the command like `mlflow models build-docker <uri> --enable-mlserver --mlserver-model-name foo`
- Pick up the `model uuid` from the `MLmodel` file of the model artifact, and use that as a model name
- Sanitize the model uri and use that as a model name (probably the least effort option)

What do you think?

### Motivation

> #### What is the use case for this feature?

When we make predictions, it is very important to trace which model was responsible for which predictions. We want to pick this metadata up from the `/v2/{model}/metadata` endpoint of mlserver. This allows us to get some reference to the model without breaking the current API contract.

> #### Why is this use case valuable to support for MLflow users in general?

In many environments such as fintech and health, traceability is not only very useful and helps with debugging, but also required by regulations.

> #### Why is this use case valuable to support for your project(s) or organization?

We need to be able to enrich inferences with some metadata about which model gave which prediction to

- Comply with traceability requirements in our field
- Compare models head-to-head in production at the same time
- AB testing (compare base model to best model)

> #### Why is it currently difficult to achieve this use case?

The service does not expose any reference to the model it is currently serving.

### Details

Suggested options in the proposal summary.

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by BenWilson2** [comment]
@dingobar thank you for the great suggestion! We're going to discuss with some of the other maintainers and get back to you on this.

**Comment by BenWilson2** [comment]
Hi @dingobar there is a PR under review by @tomasatdatabricks that will greatly help in enabling this FR:
https://github.com/mlflow/mlflow/pull/6575

Once this PR is merged, let's see about getting artifact information fed to the response object of the endpoint. Once again, thank you for bringing this up!

**Comment by dingobar** [comment]
Sounds great, having that json wrapper opens for some customization. Is that pr related to the mlserver serving, or just the native mlflow server?

I opened this issue in mlserver: https://github.com/SeldonIO/MLServer/issues/709

**Comment by tomasatdatabricks** [comment]
@BenWilson2 @dingobar  Note that the PR 6575 is against 2.0 branch - it will be released with mlflow 2.0. and we're not planning on porting it to mlflow 1.x. MLServer will need some minor changes in order to be compatible with MLflow 2.0.

**Comment by dingobar** [comment]
I have an implementation light suggestion to this: Can we set the `version` parameter for the mlserver model to the `run_id` somehow? What way, we can have a constant model name, but different versions, and request the model metadata slightly before/after inference. It still has the issue that we don't know if the model changed between the two API-calls, but it's better than nothing.

**Comment by B-Step62** [comment]
MLflow's MLServer-based serving integration is deprecated, so exposing the model id from its `/inference` endpoint is no longer relevant. Closing as not planned.
